docs/multiarch.rst

```diff
@@ -324,6 +324,122 @@ platforms will continue. Once all have either succeeded or failed, the
+a new osbs-client API method.
+
+Multiplexing
+~~~~~~~~~~~~
+
+In order to allow the client to de-multiplex logs containing a mixture
+of logs from an orchestrator build and from its worker builds, a new
+logging field, platform, is used. Within atomic-reactor all logging
+should be done through a LoggerAdapter which adds this 'platform'
+keyword to the 'extra' dict passed into logging calls. These objects
+should come from a factory function::
+
+  def get_logger(name, platform=None):
+      return logging.LoggerAdapter(logging.getLogger(name),
+                                   {'platform': platform or '-'})
+
+The logging format will include this new field::
+
+  %(asctime)s platform:%(platform)s - %(name)s - %(levelname)s - %(message)s
+
+resulting in log output like this::
+
+  2017-06-23 17:18:41,791 platform:- - atomic_reactor.foo - DEBUG - this is from the orchestrator build
+  2017-06-23 17:18:41,791 platform:x86_64 - atomic_reactor.foo -  DEBUG - text:2017-06-23 17:18:41,400 platform:- atomic_reactor.foo -  DEBUG - this is from a worker build
+
+Demultiplexing will be possible using a new osbs-client API method,
+get_orchestrator_build_logs(). This method is a generator function
+that returns objects with these attributes:
+
+platform
+  str, platform name if worker build, else None
+
+line
+  str, log line (Unicode)
+
+In order to do this it should find the third space-separated field
+from the log line. Since the asctime value contains a space between
+the date and time, the third field is for the platform.
+
+If the platform field does not start "platform:", then this is a log
+line from the orchestrator build which has (mistakenly) not been
+logged using the adapter.
+
+If the platform field matches "platform:-", then this is a log line
+from the orchestrator build.
+
+Otherwise, the platform name can be found by removing the "platform:"
+prefix.
+
+For orchestrator build logs, the line is returned as-is.
+
+For worker build logs, the worker log line is returned but with the
+third field removed.
+
+Encoding issues
+~~~~~~~~~~~~~~~
+
+When retrieving logs from containers, the text encoding used is only
+known to the container. It may be based on environment variables
+within that container; it may be hard-coded; it may be influenced by
+some other factor. For this reason, container logs are treated as byte
+streams.
+
+This applies to:
+
+- containers used to construct the built image
+- the builder image running atomic-reactor for a worker build
+- the builder image running atomic-reactor for an orchestrator build
+
+When retrieving logs from a build, OpenShift cannot say which encoding
+was used. However, atomic-reactor can define its own output encoding
+to be UTF-8. By doing this, all its log output will be in a known
+encoding, allowing osbs-client to decode it.
+
+In this way, the osbs-client get_build_logs() method will once again
+return an iterable of decoded strings, rather than of a bytes type.
```

Comment:
To do this safely across our toolchain, do we need some kind of versioning for this api method?
I'm thinking that once this is done, koji-containerbuild will need to be updated to properly handle it.